### PR TITLE
10-get-api-reviews-queries

### PR DESCRIPTION
**Pull request by alxjnr**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -84,7 +84,82 @@ describe("API Testing", () => {
             });
           });
       });
-
+      test("Should return reviews by selected category (social deduction) if specfied", () => {
+        return request(app)
+          .get("/api/reviews?category=social_deduction")
+          .expect(200)
+          .then((res) => {
+            res.body.reviews.forEach((obj) => {
```
Comment (corkesw) [inline __tests__/app.test.js:92]: As we are asserting in a forEach, you *must* check that the array length is truthy

File: __tests__/app.test.js [hunk 2]
```diff
@@ -84,7 +84,82 @@ describe("API Testing", () => {
...
+              expect(obj).toHaveProperty("category", "dexterity");
+            });
+          });
+      });
+      test("Should return a 404 if category specified does not exist", () => {
+        return request(app).get("/api/reviews?category=test").expect(404);
+      });
```
Comment (corkesw) [inline __tests__/app.test.js:109]: You should add a test for category that exists but which has no associated articles (eg children's games) which should return an empty array

File: __tests__/app.test.js [hunk 3]
```diff
@@ -84,7 +84,82 @@ describe("API Testing", () => {
...
+              expect(obj).toHaveProperty("category", "social deduction");
+            });
+          });
+      });
+      test("Should return a 400 if query values are of an invalid type", () => {
+        return request(app)
+          .get("/api/reviews?category=social_deduction&sort_by=votes&order=ooo")
+          .expect(400);
+      });
+      test("Should return a 400 if query keys are of an invalid type", () => {
```
Comment (corkesw) [inline __tests__/app.test.js]: Maybe I can't spot it but this looks like the same request as the last test (correct me if I'm wrong!) The last test checks for an invalid sort_by query. Should this be checking for an invalid order query? eg `order=bananas`

File: models/models.reviews.js [hunk 4]
```diff
@@ -15,18 +15,32 @@ exports.fetchReviewById = (review_id) => {
...
+  if (category.length) {
+    categoryAdd = `WHERE category LIKE '%${category}%'`;
+  }
+
+  let queryStr = `SELECT reviews.*, COUNT(comments.review_id) AS comment_count FROM reviews 
+  LEFT JOIN comments 
+  ON reviews.review_id = comments.review_id ${categoryAdd}
+  GROUP BY reviews.review_id 
+  ORDER BY reviews.${sort_by} ${order};`;
+
```
Comment (corkesw) [inline models/models.reviews.js:60]: There is no protection against SQL injection here so you will need some logic to check that these queries are all valid and acceptable. If they are not, then you will need to reject the Promise and handle the custom error

**Review by corkesw — COMMENTED** [review]
Great work here, the testing is really thorough, just one question on one of the tests. We don't have any protection against injection here so I would suggest adding some logic to check that the sort and order and are valid. The category can be parameterised.

File: __tests__/app.test.js (hunk 3 above)
Comment (alxjnr) [reply __tests__/app.test.js]: order here = ooo which is an inavlid order query :) will double check this though

Comment (alxjnr) [reply __tests__/app.test.js]: it is the same test, good spot

**Comment by alxjnr** [comment]
Not sure about this one. It works and is passed new tests, but I dont like that I have hard-coded categories and review columns. I didn't want to make more calls to the database than what was necessary, but adding new columns to these database would mean having to edit the model

Spent a long time trying to figure out parameterised queries and why they wouldn't work for SORT_BY and ORDER, but learnt a lot

File: models/models.reviews.js [hunk 5]
```diff
@@ -20,25 +20,60 @@ exports.fetchReviews = (
...
 
-  if (category.length) {
-    categoryAdd = `WHERE category LIKE '%${category}%'`;
+  const categoryList = [
+    "",
+    "euro game",
+    "social deduction",
+    "dexterity",
+    "childrens games",
+  ];
```
Comment (robert112233) [inline models/models.reviews.js]: I would be hesitant to hard code these as you are using different devData, my first thought would be to pull the categories from the database and make sure you get a match. There might be an easier way but I will have to get back to you!

**Review by robert112233 — COMMENTED** [review]
Really good work, the extra test looks good. I know you don't like the idea of hardCoding the categories, (I hard-coded mine on the course). I think another DB query may be necessary, I will speak to Stephen though!

**Comment by alxjnr** [comment]
Have updated the function to make calls to the db to fetch lists for validation.

**Review by Jodiech — APPROVED** [review]
Happy for you to merge 😄

**Pull request merged**